libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity/Scripts/PeerConnection.cs

```diff
@@ -1,608 +1,1208 @@
+        /// </summary>
+        public MediaKind Kind => _kind;
+
+        /// <summary>
+        /// Media sender producing the media to send through the transceiver attached to this media line.
+        /// This must be an instance of a class derived from <see cref="AudioSender"/> or <see cref="VideoSender"/>
+        /// depending on <see cref="Kind"/>.
+        /// If this is non-<c>null</c> then the peer connection will negotiate sending some media, otherwise
+        /// it will signal the remote peer that it does not wish to send (receive-only or inactive).
+        /// </summary>
+        public MediaSender Sender;
+
+        /// <summary>
+        /// Media receiver consuming the media received through the transceiver attached to this media line.
+        /// This must be an instance of a class derived from <see cref="AudioReceiver"/> or <see cref="VideoReceiver"/>
+        /// depending on <see cref="Kind"/>.
+        /// If this is non-<c>null</c> then the peer connection will negotiate receiving some media, otherwise
+        /// it will signal the remote peer that it does not wish to receive (send-only or inactive).
+        /// </summary>
+        public MediaReceiver Receiver;
+
+        /// <summary>
+        /// Transceiver attached with this media line.
+        /// On the offering peer this changes during <see cref="PeerConnection.CreateOffer"/>, while this is updated by
+        /// <see cref="PeerConnection.SetRemoteDescriptionAsync(string, string)"/> when receiving an offer on the answering peer.
+        /// </summary>
+        public Transceiver Transceiver { get; private set; }
+
+        /// <summary>
+        /// Sender actually attached during <see cref="PeerConnection.SetRemoteDescriptionAsync(string, string)"/>.
+        /// This is different from <see cref="Sender"/> until a negotiation is achieved.
+        /// </summary>
+        [NonSerialized]
+        private MediaSender _attachedSender;
+
+        /// <summary>
+        /// Receiver actually paired during <see cref="PeerConnection.SetRemoteDescriptionAsync(string, string)"/>.
+        /// This is different from <see cref="Receiver"/> until a negotiation is achieved.
+        /// </summary>
+        [NonSerialized]
+        private MediaReceiver _pairedReceiver;
+
+        public MediaLine(MediaKind kind)
+        {
+            _kind = kind;
+        }
+
+        internal async Task AttachTrackAsync()
+        {
+            Debug.Assert(Transceiver != null);
+            Debug.Assert(Sender != null);
+            Debug.Assert(_attachedSender == null);
+            if (Kind == MediaKind.Audio)
+            {
+                var audioSender = (AudioSender)Sender;
+                await audioSender.AttachTrackAsync();
+            }
+            else if (Kind == MediaKind.Video)
+            {
+                var videoSender = (VideoSender)Sender;
+                await videoSender.AttachTrackAsync();
+            }
```

Comment:
AttachTrackAsync and DetachTrack could be virtual methods in MediaSender and we'd avoid dispatching explicitly here.

Reply:
Same with OnPaired/OnUnpaired (could take a MediaTrack and cast to the specific track type internally).

Reply:
But `MediaSender` is for local (sender) and remote (receiver) tracks, while `AttachTrackAsync()` is for senders only. I wanted to avoid mixing the 2 and have some empty implementation on the receivers, which sounds weird. Same for `OnPaired()`, this is only for receivers, not senders.

Reply:
Fixed.